src/pkg_manager.ts

```diff
@@ -16,17 +18,51 @@ function modeToFlag(mode: InstallOptions["mode"]): string {
     : "";
 }
 
+function modeToFlagYarn(mode: InstallOptions["mode"]): string {
+  return mode === "dev" ? "--dev" : mode === "optional" ? "--optional" : "";
+}
+
 function toPackageArgs(pkgs: JsrPackage[]): string[] {
   return pkgs.map(
     (pkg) => `@${pkg.scope}/${pkg.name}@npm:${pkg.toNpmPackage()}`,
   );
 }
 
+async function isYarnBerry(cwd: string) {
+  // this command works for both yarn classic and berry
+  const version = await exec("yarn", ["--version"], cwd, undefined, true);
+  if (!version) {
+    logDebug("Unable to detect yarn version, assuming classic");
+    return false;
+  }
+  if (version.startsWith("1.")) {
+    logDebug("Detected yarn classic from version");
+    return false;
+  }
+  logDebug("Detected yarn berry from version");
+  return true;
+}
+
+async function getLatestPackageVersion(pkg: JsrPackage) {
+  const url = `${JSR_URL}/${pkg}/meta.json`;
+  const res = await fetch(url);
+  const body = await res.text();
+  if (!res.ok) {
+    throw new Error(`Received ${res.status} from ${url}`);
+  }
+  const { latest } = JSON.parse(body);
+  if (!latest) {
+    throw new Error(`Unable to find latest version of ${pkg}`);
+  }
```

Comment:
Nit: The `Response` class has a native `.json()` method, so we don't need to do the parsing ourselves. See https://developer.mozilla.org/en-US/docs/Web/API/Response/json

```suggestion
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`Received ${res.status} from ${url}`);
  }
  const { latest } = await res.json();
  if (!latest) {
    throw new Error(`Unable to find latest version of ${pkg}`);
  }
```

Reply:
i did this intentionally to avoid this problem in node: https://undici.nodejs.org/#/?id=garbage-collection

Reply:
Right in that case let's call `res.body?.cancel()` in the error case when checking `if (!res.ok)`

Reply:
done :pray: